### DISCUSSION TITLE
How to write tests for formulae without any scriptable interactions?

### DISCUSSION THREAD

**Discussion by superatomic** [Homebrew/discussions#3377] [Writing Formulae/Casks] [answered]
I am currently writing a formula for [the TUI todo manager `dooit`](https://github.com/kraanzu/dooit) and I have ran into a problem with adding tests. I have looked over [the guidelines for adding tests to formulae](https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula) and I am unable to find any way to test `dooit`. The application is only an interactive TUI app and does not have any scriptable interface, nor does it have any way of displaying help or version information.

The app also runs until `ctrl + Q` is inputted, which makes it so I can't even check to see if it can open and close with a non-zero status code, because it will block forever if the test succeeds.

Can I just have the test run `system "true"` and leave a comment explaining the situation, or is there something that I can do to actually have a test?

**Comment by gromgit** [comment, marked as answer]
`expect` is your friend for testing TUI applications: https://core.tcl-lang.org/expect/index. It's used in the test blocks of several dozen core formulae; I personally added such a test block to the `angband` formula recently, so you can use that as an example.

One issue with testing TUI applications is that the `expect` command to match specific application output (also called `expect`, naturally), usually doesn't work at all, due to some unfortunate interaction with ncurses. That's why the `angband` test script is peppered with `sleep` commands instead, to pace the keystrokes being sent to the running `angband`.

**Reply by carlocab** [reply]
You might not even need `expect`. In many cases, Ruby's `PTY` module does enough.

**Reply by superatomic** [reply]
Thank you for your responses. How would the `PTY` module work in this case?

**Reply by gromgit** [reply]
Check out the `amp` formula for a worked example.

**Reply by superatomic** [reply]
Thank you so much! You've both been very helpful.
